### ISSUE TITLE
[HOLD for payment 2023-10-09] HIGH: LHN logic is broken for paid reports

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#26669]
Correct logic as follows:

**`policyExpenseChat` (member view)**
 - First line: `%workspaceName%` (note: name depends on who’s looking at it, this is the member case)
 - Second line: latest message in the workspace chat. We a report has just paid paid, that's `%workspaceName% paid %amount%`

**Report (member & admin)**
 - First line: report name (e.g. `%workspaceName% paid %amount%`)
 - Second line: latest message, (e.g. `%userDisplayName% paid %amount%`

At the moment the `policyExpenseChat` is incorrectly showing the second line from the report (e.g. Niki paid...) instead of what should should actually be the second line for the `policyExpenseChat`. 

Example
<img width="270" alt="Screenshot 2023-09-04 at 09 01 56" src="https://github.com/Expensify/App/assets/1127863/56958505-6218-4596-b148-a40540560b27">

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0159fb834042f1a338</li>
        <li>Upwork Job ID: 1698698179177246720</li>
        <li>Last Price Increase: 2023-09-04</li>
    </ul>
</details>

**Comment by JmillsExpensify** [comment]
Taking this as part of waves

**Comment by JmillsExpensify** [comment]
Adding this to wave4 so that we ensure it gets picked up.

**Comment by mountiny** [comment]
Closed https://github.com/Expensify/App/issues/26542 in favour of this one, posting my findings from there here as well:

____

I have tried to get to the bottom of this but its getting quite convoluted with all the perspectives and if user is owed  or its settled. Its doable just the logic is getting too tricky and its prone to cause some regressions in future. Might have to clean this up a bit.

I will move this to weekly and will look into it later.

Maybe @koko57 or @BeeMargarida would be interested in taking this on.

Currently the [getReportPreviewMessage](https://github.com/Expensify/App/blob/393475bd0e41b533352caf254550fc6124909b30/src/libs/ReportUtils.js#L1424-L1469) for both Money report and the workspace or IOU chat last message in LHN. But its misleading and I think we should separate the logic of the report preview action and the paid report action message into two functions

Then I think we should simplify the translations that in case of expense report, the paid message will always include the Policy name same as in case of `Workspace owes £40` we would have `Workspace paid £40`. Then we need to update the logic that adds the last actor to not add it for this kind of message in LHN

**Comment by mountiny** [comment]
This does not have to be internal, but he logic is quite complicated at this point and we are reusing the same getReportPreviewMessage method to get the message for report preview and also for the money report, which was fine until we have added more than an admin role and I think we need to separate this. 

I think it would be a good issue for some agency

**Comment by koko57** [comment]
@mountiny I'm currently working on another ticket but as I'm waiting for a PR review I can have a look 🙂

**Comment by mountiny** [comment]
🎉 lets do this! 

Let us know if things are not clear from the issue body?

I think we will have to split the `getReportPreviewMessage` into two methods to make this cleaner even though they share lots of similar logic

**Comment by koko57** [comment]
Either I'm doing something wrong or it's fixed - I cannot recreate the issue (both prod and staging).
Could you please provide some repro steps? 

![Screenshot 2023-09-14 at 10 10 48](https://github.com/Expensify/App/assets/28020445/ff7eb88a-5c82-40ea-9ee6-658e9a3ad002)

**Comment by koko57** [comment]
@mountiny @JmillsExpensify 
So just to clarify:
in this case 
![Screenshot 2023-09-14 at 10 16 04](https://github.com/Expensify/App/assets/28020445/a231fcdf-0d99-45fc-80fa-cca130af06c0)

the second line should be: "TEST paid 3.00 elsewhere"?

**Comment by mountiny** [comment]
cc @JmillsExpensify 

Correct, In this case there will be the workspace name twice

**Comment by koko57** [comment]
@mountiny I've found the culprit here: 
the name displayed for the member was created by this line of code https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/SidebarUtils.js#L291-L292

and then added to the lastMessageTextFromReport here https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/SidebarUtils.js#L293 (btw do we need this condition with report? the function returns when there is no report)

but as we don't add the payer to the string like here https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/ReportUtils.js#L1467
but we pass only amount 
https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/ReportUtils.js#L1459

we end up with incomplete string for the workspace admin
![Screenshot 2023-09-14 at 14 47 37](https://github.com/Expensify/App/assets/28020445/8cb3cc94-c86d-4bc3-a38d-d130a5695a7e)
and the string with added name for the member

![Screenshot 2023-09-14 at 14 56 30](https://github.com/Expensify/App/assets/28020445/af5d8f75-37cb-4c51-bf37-24e8b71d960b)

Starting working on my proposal, I guess the solution is pretty easy and won't require splitting the logic of [getReportPreviewMessage](https://github.com/Expensify/App/blob/393475bd0e41b533352caf254550fc6124909b30/src/libs/ReportUtils.js#L1424-L1469) for the report and policyExpenseChat, but I need to check if it won't cause any regressions.

**Comment by koko57** [comment]
and one more question: we don't need the ":" after the workspace/payer's name?

**Comment by koko57** [comment]
and the second one:
is it ok that we don't have the payer's name here?
Here's how it currently looks like for
the requestor:
![Screenshot 2023-09-14 at 15 18 19](https://github.com/Expensify/App/assets/28020445/40807bc5-3a71-455b-b1b3-d0b1cc47ff02)
and the payer:
![Screenshot 2023-09-14 at 15 17 47](https://github.com/Expensify/App/assets/28020445/16556790-1101-4ca2-971f-5c2901c78907)

**Comment by mountiny** [comment]
> PS. I also wondered why it works properly for 'owes' action -> we pass payer's name to the string and the condition for adding the name with colon is false - the lastActorAccountId is 0. Is it correct behavior?

yes

This is looking good to me and its what I run into when investigated, however, I thought I would do the same with payer owes translation key in the getReportPreviewMesage but for the paid case and workspace too. And in such case its just getting a bit more confusing as the expense report message is not going to be the same everywhere as the report preview. So why to use the same method. 

> is it ok that we don't have the payer's name here?

I think that is fine, but @JmillsExpensify to confirm. We do post who owes.

> (btw do we need this condition with report? the function returns when there is no report)
👍 

> and one more question: we don't need the ":" after the workspace/payer's name?

We dont need it I would say

**Comment by koko57** [comment]
> PS. I also wondered why it works properly for 'owes' action -> we pass payer's name to the string and the condition for adding the name with colon is false - the lastActorAccountId is 0. Is it correct behavior?

About the correct behavior - I meant: is the fact that lastActorAccountId is 0 ok?

**Comment by JmillsExpensify** [comment]
> and one more question: we don't need the ":" after the workspace/payer's name?

This is a great question. Agree with @mountiny that we don't need it for this case.

**Comment by JmillsExpensify** [comment]
> and the second one: is it ok that we don't have the payer's name here? Here's how it currently looks like for the requestor: ![Screenshot 2023-09-14 at 15 18 19](https://user-images.githubusercontent.com/28020445/267987699-40807bc5-3a71-455b-b1b3-d0b1cc47ff02.png) and the payer: ![Screenshot 2023-09-14 at 15 17 47](https://user-images.githubusercontent.com/28020445/267987818-16556790-1101-4ca2-971f-5c2901c78907.png)
> 
> What are other possible cases where we don't want to display payer's name?

I think all these cases look fine as well. I also can't think of any other cases where we don't want to display payer's name.

**Comment by mountiny** [comment]
> is the fact that lastActorAccountId is 0 ok?

@koko57 you are saying the `report.lastActorAccountID` is 0 when the workspace admin pays? Given in the screenshot it is showing Niki, then in that case it was updated correctly

I dont think it should be 0

**Comment by koko57** [comment]
@mountiny no, it happens when we request monet in the workspace
![Screenshot 2023-09-15 at 10 49 29](https://github.com/Expensify/App/assets/28020445/70f5fbdc-f7d1-427a-a630-bf145cb573d6)
that's why it doesn't match the condition and the payer's name isn't added like here
![Screenshot 2023-09-15 at 10 50 58](https://github.com/Expensify/App/assets/28020445/4cdc4ce6-ff3e-4b46-83f3-443bd33e1ee1)

**Comment by mountiny** [comment]
@koko57 what is the report contents?

**Comment by koko57** [comment]
![Screenshot 2023-09-15 at 11 21 15](https://github.com/Expensify/App/assets/28020445/3eb51024-b1c1-4118-a3dc-1e82231b5d4d)

**Comment by mountiny** [comment]
@koko57 this is the workspace chat and in the expense report, the account ID is not 0?

**Comment by koko57** [comment]
yeah, in the expense report is not 0 🤔 
![Screenshot 2023-09-15 at 12 05 21](https://github.com/Expensify/App/assets/28020445/dd9acfdd-4894-45b3-beeb-e33757b0fddc)

**Comment by koko57** [comment]
I guess I've found a solution but I need to check if it doesn't cause any regression

**Comment by mountiny** [comment]
Hmm this is a bit odd indeed, but currently dont have much time to dig into this one on backend

**Comment by koko57** [comment]
@mountiny was about to paste my proposal 😅

**Comment by koko57** [comment]
@reviewers

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

We have an incorrect message for the policyExpenseChat when the money request is settled.
Correct messages:
First line: %workspaceName% (note: name depends on who’s looking at it, this is the member case)
Second line: latest message in the workspace chat. When a report has just paid, that's %workspaceName% paid %amount%

Current messages:
First line: %workspaceName% (note: name depends on who’s looking at it, this is the member case)
%lastActorName% paid %amount%

### What is the root cause of that problem?

The name displayed for the member was created by this line of code:
https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/SidebarUtils.js#L291-L292

and then added to the lastMessageTextFromReport here:
https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/SidebarUtils.js#L293

As we don't add the payer to the string like here:
https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/ReportUtils.js#L1467
but we pass only amount
https://github.com/Expensify/App/blob/95c9e4743f44ca42ece2c6ac56259b911d5feafc/src/libs/ReportUtils.js#L1459

we end up with incomplete string for the workspace admin ‘paid PLN 3.00 elsewhere’
and the string with added name for the member ‘Aga K: paid PLN 3.00 elsewhere’
To the other messages it didn’t add the name because either the lastActorDetails.accountID was 0 or participants count was 1.

### What changes do you think we should make in order to solve the problem?

If we want to keep the current logic of getReportPreviewMessage and not to add the payer to the settled iou we should do 2 things:

Block adding the member name to the message
we can do it by saving this [line](https://github.com/Expensify/App/blob/2427a06ed24f0649e55c14dbf4ae1a2a9ebe2109/src/libs/SidebarUtils.js#L292) as a const
```
const lastActor = hasMultipleParticipants && lastActorDetails && lastActorDetails.accountID && Number(lastActorDetails.accountID) !== currentUserAccountID ? lastActorDetails.displayName : '';
```
and then add this condition [here](https://github.com/Expensify/App/blob/2427a06ed24f0649e55c14dbf4ae1a2a9ebe2109/src/libs/SidebarUtils.js#L305-L316):
```
} else if (lodashGet(lastAction, 'actionName', '') !== CONST.REPORT.ACTIONS.TYPE.REPORTPREVIEW && lastActor && lastMessageTextFromReport) {
result.alternateText = `${lastActor}: ${lastMessageText}`;
```
2.  Add workspace name to the message
by adding it [here](https://github.com/Expensify/App/blob/83b3f5d501559840f1a773ee76618625e4dbabe6/src/libs/OptionsListUtils.js#L398-L400):
```
const actorName = ReportUtils.isSettled(iouReport.reportID) ? `${ReportUtils.getPolicyName(report)} ` : ``
lastMessageTextFromReport = actorName + ReportUtils.getReportPreviewMessage(iouReport, lastReportAction);
```
this way we only limit the changes to the getLastMessageTextForReport and getOptionData without changing the logic of getReportPreviewMessage.

### What alternative solutions did you explore? (Optional)

Currently we don’t have the payer’s name in any of the chat. For consistency maybe we could add a payer’s name to the message like we do for the other messages (like those with ‘owes’). We would also need to apply the first condition.

I’d rather not create a separate method for getting preview message just for this one case - it would require duplicating code and adding additional conditions to either use one method or the latter.

**Comment by mountiny** [comment]
This looks good to me, I think I would name the `lastActor` as `lastActorDisplayName` to be more specific. Can you please raise a PR? thanks!

**Comment by koko57** [comment]
@mountiny I've opened the draft PR https://github.com/Expensify/App/pull/27700 

but one last thing, just to make sure:

for the money request report it should stay as it is?
![Screenshot 2023-09-18 at 19 05 07](https://github.com/Expensify/App/assets/28020445/5c8c9e7c-41f6-4a2c-8787-28f6ed668a64)

I mean without the workspace name? Or should it also have the workspace name? (if so it's not a big change, one-liner)

**Comment by mountiny** [comment]
Correct, for the money request report it should be without the name. Can we also update IOUTests to cover this by any chance?

**Comment by koko57** [comment]
@mountiny Could it be done in a follow-up? Btw as it's only affecting the Sidebar and Option Row (SidebarUtils and OptionListUtils) does it really have to be covered in IOUTests?

**Comment by mountiny** [comment]
It can be a follow up, the tests should make sure the message is correctly composed for the LHN, but yeah it might be better for different test file

**Comment by koko57** [comment]
PR opened for the review https://github.com/Expensify/App/pull/27700

**Comment by dylanexpensify** [comment]
@mananjadhav can we please get the above reviewed? TY!

**Comment by JmillsExpensify** [comment]
@mananjadhav bump

**Comment by JmillsExpensify** [comment]
Ah, actually I see [your response](https://github.com/Expensify/App/pull/27700#issuecomment-1734463695) a couple of hours ago in the PR. Thanks, let's stay focused on this as we'd like to close the loop on this issue this week.

**Comment by mananjadhav** [comment]
I'll be on top of this one.

**Comment by dylanexpensify** [comment]
@mananjadhav bump on where we're at here?

**Comment by dylanexpensify** [comment]
Merged and on staging!

**Comment by JmillsExpensify** [comment]
@mananjadhav Do you mind kicking off the BZ checklist?

**Comment by JmillsExpensify** [comment]
Payment summary: $500 to @mananjadhav based on PR review.

**Comment by JmillsExpensify** [comment]
Still waiting on the checklist and then we can close this out.

**Comment by mananjadhav** [comment]
I am not sure if we should treat this as a regression as I couldn't find a PR where we had this messaging requirement. @mountiny do you have that? Hence, I don't have a specific offending PR.

But I do feel we should add a regression test. QA Steps from the PR look good to me.

**Comment by mountiny** [comment]
@mananjadhav I agree finding the regression for this can be tricky due to complexity of the LHN code

Can you comment the steps here so its easier for the bugZero assignee to handle? Thanks!

**Comment by trjExpensify** [comment]
👋 Did the work for this issue break the iou/expenseReport preview logic? 

<img width="306" alt="image" src="https://github.com/Expensify/App/assets/16232057/692f21f9-92ea-4572-b924-8a7a339da291">

It should be:

>[%displayName% || %workspaceName%] owes £21.96
>%firstName%: requested £5.00 for [%merchant% || %comment%]

I.e 

**IOU Report**
>Jeff Banks owes £21.96
>Tom: requested £5.00 for Thursday Lunch

**Expense Report**
>Duraflame owes £21.96
>Tom: requested £5.00 for Home Depot

**Comment by JmillsExpensify** [comment]
Great reason that we need a regression test for this case. @mananjadhav do you mind suggesting one please?

**Comment by trjExpensify** [comment]
Hm, and another one. Why isn't this reportAction coming from the reimburser who clicked `Pay`? We're now using the workspace name, so that means there's no audit history of who actually paid an expenseReport on behalf of the company. That's not great for compliance, we need to fix that as well please! 

<img width="1434" alt="image" src="https://github.com/Expensify/App/assets/16232057/8b62fcfa-58e9-4aab-b210-e7dc68ed776b">

**Comment by mananjadhav** [comment]
*Regression Test proposal:*

1. Login to the app in two devices - one as a workspace member (assume userA) and the other as the owner (ownerB).
2. As the member, request money (assume $50) from the workspace (example controlRoom)
3. As the owner pay for the request.
4. As the member you should see the policy expense chat in LHS as well as preview as:
	`<workspace-name> paid <amount> <payment-method>`

    With our example it would:
		`controlRoom paid $50 elsehwere`


@mountiny @JmillsExpensify ^

Also should we create a new issue for @trjExpensify's [comment](https://github.com/Expensify/App/issues/26669#issuecomment-1761328482) ?

**Comment by mountiny** [comment]
I agree with Manan that this PR did not touch the logic of who [the action is coming from](https://github.com/Expensify/App/issues/26669#issuecomment-1761328482).

And it feels like the subscript in of the paid expense report, that could be mixture of the backend changes to return both participants in the expense report and this change. 

Either way we made this logic exteremely brittle depending of various things and these never ending regressions are consequence of it. I am not sure how to tackle this. 

@koko57 we could try to abstract away some of the logic to determine to LHN preview messages and then create unit tests for that method based on various scenarios of the reports, would that be feasible?

**Comment by koko57** [comment]
@trjExpensify on your comment: https://github.com/Expensify/App/issues/26669#issuecomment-1760229054 could you please explain what is wrong here? My PR is only adding "[LAST_ACTOR_NAME]:" where necessary (it was originally also adding the workspace name for the workspace requests https://github.com/Expensify/App/pull/27700#issuecomment-1734999067).

@mountiny @trjExpensify So we want to display a different message in the chat (with the payer's name) and the workspace name in the sidebar?

**Comment by mountiny** [comment]
@koko57 I dont think this comes from your PR as it tested well there, there must have been some other change. However, if you have time to investigate that would be helpful

And yeah you are correct, in the LHN we should show the Workspace paid, but in the expense report itself for auditing purposes we should show which person actually paid cc @trjExpensify

**Comment by trjExpensify** [comment]
>And yeah you are correct, in the LHN we should show the Workspace paid,

You're talking about the title of the expenseReport in the LHN, yeah?

**Comment by mountiny** [comment]
<img width="177" alt="image" src="https://github.com/Expensify/App/assets/36083550/41595911-cdf0-4ba2-8ae7-75f2efbb49a9">

Both of these

**Comment by trjExpensify** [comment]
Ah, so no.. the preview is of the report action added to the report and should come from whichever user took the action. 

When someone pays an expenseReport:

1. The reportAction added to the report comes from the reimburser who paid
2. The LHN title updates to "%workspaceName% paid %reportTotal%"
3. The LHN preview shows (1). 

I.e 

**Report Action**
**Tom Rhys Jones 11:58am**
paid $50.00 elsewhere

**LHN**
Duraflame paid $50.00
Tom: paid $50.00 elsewhere

**Comment by mountiny** [comment]
Ah I missed it with the workspace chat preview 
<img width="185" alt="image" src="https://github.com/Expensify/App/assets/36083550/3001012a-8006-4d67-9d53-447bcfc32e55">

Then there its correctly saying the `%workspaceName% paid this`

**Comment by JmillsExpensify** [comment]
Shall we create a fresh issue then? For what it's worth, I like this idea from @mountiny:

> @koko57 we could try to abstract away some of the logic to determine to LHN preview messages and then create unit tests for that method based on various scenarios of the reports, would that be feasible?

**Comment by JmillsExpensify** [comment]
P.S. Regression test has been added.

**Comment by JmillsExpensify** [comment]
$500 payment approved for @mananjadhav based on [this summary](https://github.com/Expensify/App/issues/26669#issuecomment-1752029327).

**Comment by koko57** [comment]
@JmillsExpensify @mountiny should I start working on this one? 🙂 Will a new ticket be created?

**Comment by mananjadhav** [comment]
New issue would be helpful and I can help with the review.

**Comment by mountiny** [comment]
Yes @trjExpensify would you be able to create a new one for this? You will be the best in defining exactly what needs to be done

**Comment by trjExpensify** [comment]
@koko57 can you write a message on the new issue so I can assign you? Thanks! https://github.com/Expensify/App/issues/30042

**Comment by mountiny** [comment]
Thanks this can be closed then ❤️
